google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/OpenCensusUtil.java

```diff
@@ -0,0 +1,197 @@
+
+import io.opencensus.common.Scope;
+import io.opencensus.tags.propagation.TagContextBinarySerializer;
+import io.opencensus.tags.propagation.TagContextDeserializationException;
+import io.opencensus.tags.propagation.TagContextSerializationException;
+import io.opencensus.tags.TagContext;
+import io.opencensus.tags.Tagger;
+import io.opencensus.tags.Tags;
+import io.opencensus.trace.Link;
+import io.opencensus.trace.Span;
+import io.opencensus.trace.SpanId;
+import io.opencensus.trace.SpanContext;
+import io.opencensus.trace.TraceId;
+import io.opencensus.trace.TraceOptions;
+import io.opencensus.trace.Tracer;
+import io.opencensus.trace.Tracing;
+import io.opencensus.trace.samplers.Samplers;
+
+import java.util.Base64;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Utilities for propagating OpenCensus {@link TagContext} and {@link SpanContext} from publishers
+ * to subscribers.
+ */
+final class OpenCensusUtil {
+  private static final Logger logger = Logger.getLogger(OpenCensusUtil.class.getName());
+
+  public static final String TAG_CONTEXT_KEY = "googclient_OpenCensusTagContextKey";
+  public static final String TRACE_CONTEXT_KEY = "googclient_OpenCensusTraceContextKey";
+
+  private static final Tagger tagger = Tags.getTagger();
+  private static final TagContextBinarySerializer serializer =
+      Tags.getTagPropagationComponent().getBinarySerializer();
+
+  private static final TraceOptions SAMPLED = TraceOptions.builder().setIsSampled(true).build();
+  private static final Tracer tracer = Tracing.getTracer();
+
+  // Used in Publisher.
+  // TODO: consider adding configuration support to control adding these attributes.
+  static PubsubMessage putOpenCensusAttributes(PubsubMessage message) {
+    PubsubMessage.Builder builder = PubsubMessage.newBuilder(message);
+    String encodedSpanContext = encodeSpanContext(tracer.getCurrentSpan().getContext());
+    String encodedTagContext = encodeTagContext(tagger.getCurrentTagContext());
+    if (encodedSpanContext.isEmpty() && encodedTagContext.isEmpty()) {
+      return message;
+    }
+    if (!encodedSpanContext.isEmpty()) {
+      builder.putAttributes(TRACE_CONTEXT_KEY, encodedSpanContext);
+    }
+    if (!encodedTagContext.isEmpty()) {
+      builder.putAttributes(TAG_CONTEXT_KEY, encodedTagContext);
+    }
+    return builder.build();
+  }
+
+  // Used in Subscriber.
+  static MessageReceiver createOpenCensusMessageReceiver(MessageReceiver receiver) {
+    return new OpenCensusMessageReceiver(receiver);
+  }
+
+  private static String encodeSpanContext(SpanContext ctxt) {
+    TraceId traceId = ctxt.getTraceId();
+    SpanId spanId = ctxt.getSpanId();
+    TraceOptions traceOpts = ctxt.getTraceOptions();
+    if (traceOpts.isSampled()) {
+      return "";
+    }
+    return "traceid=" + traceId.toLowerBase16()
+        + "&spanid=" + spanId.toLowerBase16()
+        + "&traceopt=" + (traceOpts.isSampled() ? "t&" : "f&");
```

Comment:
should this line now be `"&traceopt=f&"`?

Reply:
I've changed the text encoding of span to use our standard TraceContext text format.